### ISSUE TITLE
Passing variable to page.$$eval is not working

### ISSUE THREAD

**Issue by TCPPERFUSER** [puppeteer/puppeteer#5165]
### Steps to reproduce
Run the script with the hard coded selector works.
Run the script with the selector variable gives an error
**Tell us about your environment:**

* Puppeteer version:
* Platform / OS version:
* URLs (if applicable):
* Node.js version:

**What steps will reproduce the problem?**

_Please include code that reproduces the issue._

const puppeteer = require('puppeteer');
(async () => {
    const browser = await puppeteer.launch({headless: false});
    const page = await browser.newPage();
    await page.goto("https://www.childrensplace.com/us/home", { waitUntil: 'networkidle0' });

    let topNav = await page.$$eval('li[class="navigation-level-one"]',  e => {
        var topNavCount = document.querySelectorAll('.navigation-level-one').length;
        var index = Math.floor(Math.random() * Math.floor(topNavCount));
        return "#" + e[index].id 
    });
//    console.log("topNav :" + topNav );  
    
//  Click Category
//    await pageAction(page, "Click Category", "click", "", topNav );
//  Hard Coded Category
    topNav = '#gn-level-one-li-47501';
    await page.click(topNav, { waitUntil: 'networkidle0' }).catch(err => console.log(err)); //https://www.childrensplace.com/us/c/toddler-boy-clothes

//---------------------- Start subCategory

    var selector = topNav + ' > div > div.sub-menu > div > div:nth-child(1) > div > ol > li[class="sub-menu-category-item"]';
    var selector = '#gn-level-one-li-47501 > div > div.sub-menu > div > div:nth-child(1) > div > ol > li[class="sub-menu-category-item"]';
    console.log("selector : " +  selector );  
    let subCategoryA = await page.$$eval(selector,  e => {
//        var count = document.querySelectorAll('#gn-level-one-li-47501 > div > div.sub-menu > div > div:nth-child(1) > div > ol > li[class="sub-menu-category-item"]').length;
        var count = document.querySelectorAll(selector).length;
        var index = Math.floor(Math.random() * Math.floor(count));
        return "#" + e[index].id 
    }, selector);

    console.log("subCategoryA : " +  subCategoryA );  

})();

function getRandomInt(max) {
    return Math.floor(Math.random() * Math.floor(max));
}
**What is the expected result?**
Should not get an error and gets the returned value

**What happens instead?**
Getting an error:
(node:9300) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by kblok** [comment]
Could you improve the code formatting in the description?
It's hard to read in this way.

**Comment by stevepiron** [comment]
I'm having the same issue with Mac OS X 10.15.3, Node 8.9.4 or 12.15.0 (I use both versions) and Puppeteer 2.1.1.

## Option 1: environment variable

`USER=myusername node script.js`

```node
const puppeteer = require('puppeteer');

const ENV_USER = process.env.USER;
console.log(process.env.ENV_USER, ENV_USER); // OK

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  await page.goto('https://...', {waitUntil: "networkidle2"});
  await page.$eval('#USERNAME', el => el.value = ENV_USER); // Not OK

  // ...
})();
```

## Option 2: process variable

`node script.js myusername`

```node
const puppeteer = require('puppeteer');

const USER = process.argv[2];
console.log(process.argv[2], USER); // OK

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  await page.goto('https://...', {waitUntil: "networkidle2"});
  await page.$eval('#USERNAME', el => el.value = USER); // Not OK

  // ...
})();
```

## Option 3: variable in the script

`node script.js`

```node
const puppeteer = require('puppeteer');

const USER = 'myusername';
console.log(USER); // OK

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  await page.goto('https://...', {waitUntil: "networkidle2"});
  await page.$eval('#USERNAME', el => el.value = USER); // Not OK

  // ...
})();
```

**Comment by kblok** [comment]
@stevepiron what are you getting?

**Comment by stevepiron** [comment]
Hi @kblok here's the log:

```
(node:87026) UnhandledPromiseRejectionWarning: Error: Evaluation failed: ReferenceError: USER is not defined
    at __puppeteer_evaluation_script__:1:19
    at ExecutionContext._evaluateInternal (/Users/stevepiron/node_modules/puppeteer/lib/ExecutionContext.js:122:13)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
    at async ExecutionContext.evaluate (/Users/stevepiron/node_modules/puppeteer/lib/ExecutionContext.js:48:12)
    at async ElementHandle.evaluate (/Users/stevepiron/node_modules/puppeteer/lib/JSHandle.js:54:12)
    at async ElementHandle.$eval (/Users/stevepiron/node_modules/puppeteer/lib/JSHandle.js:505:20)
    at async /Users/stevepiron/example.js:18:3
  -- ASYNC --
    at ExecutionContext.<anonymous> (/Users/stevepiron/node_modules/puppeteer/lib/helper.js:111:15)
    at ElementHandle.evaluate (/Users/stevepiron/node_modules/puppeteer/lib/JSHandle.js:54:42)
    at ElementHandle.<anonymous> (/Users/stevepiron/node_modules/puppeteer/lib/helper.js:112:23)
    at ElementHandle.$eval (/Users/stevepiron/node_modules/puppeteer/lib/JSHandle.js:505:40)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
    at async /Users/stevepiron/example.js:18:3
  -- ASYNC --
    at ElementHandle.<anonymous> (/Users/stevepiron/node_modules/puppeteer/lib/helper.js:111:15)
    at DOMWorld.$eval (/Users/stevepiron/node_modules/puppeteer/lib/DOMWorld.js:156:21)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
    at async /Users/stevepiron/example.js:18:3
  -- ASYNC --
    at Frame.<anonymous> (/Users/stevepiron/node_modules/puppeteer/lib/helper.js:111:15)
    at Page.$eval (/Users/stevepiron/node_modules/puppeteer/lib/Page.js:345:29)
    at Page.<anonymous> (/Users/stevepiron/node_modules/puppeteer/lib/helper.js:112:23)
    at /Users/stevepiron/example.js:18:14
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
(node:87026) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:87026) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by kblok** [comment]
@stevepiron  You should do:

```js 
await page.$eval('#USERNAME', (el, user) => el.value = user, USER);
```

**Comment by stevepiron** [comment]
@kblok amazing, thanks for pointing that out. It works 🎉

**Comment by nanthanwa** [comment]
Thank you @kblok, It works.

**Comment by PrasannaBrabourame** [comment]
@kblok 
Is there possibility to use addition operator in page.eval

`var add = 0
await page.$eval('#USERNAME', (el,ad) => add += el.value ,add);
`

**Comment by kblok** [comment]
Yes @PrasannaBrabourame. `$eval` will return the result of the evaluate function.
Internally `$eval` calls `$` and then `evaluate`, returning the value of the `evaluate` function.

**Comment by eugene0928** [comment]
@kblok I have done what you showed and it works very well. I wrote e2e tests and run them via gulp with ```gulp unit``` command and they also passed. But when I try to get code coverage via gulp with ```gulp coverage```, I got internal server error and received 500 status code instead of 200.

The error message is:
```js
'Evaluation failed: ReferenceError: cov_2os11p00jv is not defined at pptr://__puppeteer_evaluation_script__:1:16'
```

And this error comes from the ```$eval()``` method where I passed variable to method as you showed.
How can I fix this?
